Add plugins guide on error handling (#26518)
grafana/grafana: docs/sources/developers/plugins/error-handling.md

+++
title = "Error handling"
type = "docs"
+++

# Error handling

This guide explains how to handle errors in plugins.

## Provide usable defaults

Allow the user to learn your plugin in small steps. By providing a usable default configuration, the user can get started right away, and you can avoid unnecessary error messages early on.

For example, by selecting the first field of a expected type, the panel can display a visualization without any user configuration. If a user explicitly selects a field we use that one, otherwise we default to the first field of type `string` that we find:

```ts
const numberField = frame.fields.find(field =>
  options.numberFieldName ? field.name === options.numberFieldName : field.type === FieldType.number
);
```

## Display errors messages

To display an error message to the user, `throw` an `Error` with the message you want to display:

```ts

> throw new Error('An error occurred.');

Reviewer: Should we really end exception messages with a period? It's normally not good style.

Reviewer 2: This really needs to be addressed in the UX style guide so that we are consistent. Hey @sarlinska, do we have a rule for this?
If not, this thread has good discussion: https://ux.stackexchange.com/questions/18671/periods-at-the-end-of-a-sentence-in-alert-message/18673

MSG says to leave periods off of UI text, which this might count as: https://docs.microsoft.com/en-us/style-guide/punctuation/periods

Reviewer: @oddlittlebird It's not a question of UX, but of programming conventions, so Microsoft's style guide and the linked StackExchange question are not relevant for the topic at hand. It is also not relevant for the UX style guide. Convention dictates that programming exceptions/errors should not be punctuated.  I couldn't find anything stating this for TypeScript, but [here](https://github.com/golang/go/wiki/CodeReviewComments#error-strings) it's stated for Go for example. For as long as I've been programming (close to 24 years), I've never seen punctuated exceptions/errors, so this is a first for me.

When it comes to UX, I think instead (as I mention in another comment) that the front-end code should be responsible for displaying a punctuated error message to the user, but that's independent of the internal exception error strings.

Reviewer: See my comment [here](https://github.com/golang/go/wiki/CodeReviewComments#error-strings). If the UI doesn't punctuate the error messages, it should be done there.

Author: Again, I agree completely with you, @aknuds1. However, this how error handling works at the moment, and the guide should reflect that. If, and when we decide to improve error handling for plugin, we'll update this guide.

Reviewer: I still think we shouldn't teach plugin authors an anti-pattern and it should be fixed in Grafana itself.

```

Grafana displays the error message in the top-left corner of the panel.

{{< docs-imagebox img="/img/docs/panel_error.png" class="docs-image--no-shadow" max-width="850px" >}}

> Avoid displaying overly technical error messages to the user. If you want to let technical users to be able to report the error, consider logging the error instead.

Reviewer: ```suggestion
Avoid displaying overly technical error messages to the user. If you want to let technical users report an error, consider logging it instead.
```

Reviewer: This still reads weird, due to the repetitive use of "the error".

```ts
try {
  failingFunction();
} catch (err) {
  console.error(err);

>   throw new Error('Something went wrong.');

Reviewer: Should we really end exception messages with a period? It's normally not good style.

Author: It's not, but the string is shown in the UI as a error message to the user, and I think UI errors should be correct sentences.

Author: I guess we should add this as a note though.

Reviewer: @marcusolsson it should be fixed in the UI, instead of requiring plugin authors to work around the issue.

Author: I agree 100%: Error handling needs to be improved for plugin development. However, this guide isn't proposing a change to how plugins handle errors, but merely documents the _current_ "best" practice. 

Regardless, after talking with UX, single sentence error messages shouldn't end with a period even in the UI, so removing the period.

Reviewer: So the UI don't punctuate the error messages for a reason? That's actually the perfect example of why error/exception strings shouldn't be punctuated, because the punctuation is entirely up to whoever displays the messages (UI is only one potential destination, typically errors are also output in logs f.ex.). Glad it's resolved.

}
```

Here are some examples of situations where you might want to display an error to the user.

### Invalid query response

Users have full freedom when they create data source queries for panels. If your panel plugin requires a specific format for the query response, use the panel canvas to guide the user.

```ts
if (!numberField) {

>   throw new Error('Query result is missing a number field.')

Reviewer: Should we really end exception messages with a period? It's normally not good style.

}

if (frame.length === 0) {

>   throw new Error('Query returned an empty result.')

Reviewer: Should we really end exception messages with a period? It's normally not good style.

+++
title = "Error handling"
type = "docs"
+++

# Error handling

This guide explains how to handle errors in plugins.

## Provide usable defaults

> Allow the user to learn your plugin in small steps. By providing a usable default configuration, the user can get started right away, and you can avoid unnecessary error messages early on.

Reviewer 2: ```suggestion
Allow the user to learn your plugin in small steps. Provide a useful default configuration so that: 
- The user can get started right away. 
- You can avoid unnecessary error messages.
```

> For example, by selecting the first field of a expected type, the panel can display a visualization without any user configuration. If a user explicitly selects a field we use that one, otherwise we default to the first field of type `string` that we find:

Reviewer 2: ```suggestion
For example, by selecting the first field of an expected type, the panel can display a visualization without any user configuration. If a user explicitly selects a field, then use that one. Otherwise, default to the first field of type `string`:
```

Reviewer 2: I am so confused by your use of "we" here. Is "we" the plugin?

```ts
const numberField = frame.fields.find(field =>
  options.numberFieldName ? field.name === options.numberFieldName : field.type === FieldType.number
);
```

> ## Display errors messages

Reviewer 2: ```suggestion
## Display error messages
```

> To display an error message to the user, `throw` an `Error` with the message you want to display:

Reviewer 2: And then it switches from "we" to "you."

```ts
throw new Error('An error occurred.');
```

Grafana displays the error message in the top-left corner of the panel.

{{< docs-imagebox img="/img/docs/panel_error.png" class="docs-image--no-shadow" max-width="850px" >}}

> Avoid displaying overly technical error messages to the user. If you want to let technical users report an error, consider logging it instead.

Reviewer 2: ```suggestion
Avoid displaying overly-technical error messages to the user. If you want to let technical users report an error, consider logging it instead.
```

```ts
try {
  failingFunction();
} catch (err) {
  console.error(err);
  throw new Error('Something went wrong.');
}
```

> > Note: Since Grafana displays the exception message in the UI as is, we recommend using grammatically correct sentences, such as ending with a period.

Reviewer 2: ```suggestion
> **Note:** Grafana displays the exception message in the UI as written, so we recommend using grammatically correct sentences. For more information, refer to the [Documentation style guide](https://github.com/grafana/grafana/blob/master/contribute/style-guides/documentation-style-guide.md).
```

Here are some examples of situations where you might want to display an error to the user.

### Invalid query response

> Users have full freedom when they create data source queries for panels. If your panel plugin requires a specific format for the query response, use the panel canvas to guide the user.

Reviewer 2: ```suggestion
Users have full freedom when they create data source queries for panels. If your panel plugin requires a specific format for the query response, then use the panel canvas to guide the user.
```
